File: src/parsing/operations/create_relationship.ts
import Database from "../../graph/database";
import Relationship from "../../graph/relationship";
import ASTNode from "../ast_node";
import Operation from "./operation";

class CreateRelationship extends Operation {
    private _relationship: Relationship | null = null;
    private _statement: ASTNode | null = null;
    private _isStatic: boolean = false;
    private _refreshEveryMs: number | null = null;
    constructor(
        relationship: Relationship,
        statement: ASTNode,
        isStatic: boolean = false,
        refreshEveryMs: number | null = null
    ) {
        super();
        this._relationship = relationship;
        this._statement = statement;
        this._isStatic = isStatic;
        this._refreshEveryMs = refreshEveryMs;
    }
    public get relationship(): Relationship | null {
        return this._relationship;
    }
    public get statement(): ASTNode | null {
        return this._statement;
    }
    public get isStatic(): boolean {
        return this._isStatic;
    }
    public get refreshEveryMs(): number | null {
        return this._refreshEveryMs;
    }
    public run(): Promise<void> {
        return new Promise(async (resolve, reject) => {
            try {
                if (this._relationship === null) {
                    throw new Error("Relationship is null");
                }
                const db = Database.getInstance();
                db.addRelationship(
                    this._relationship,
                    this._statement!,
                    this._isStatic,
                    this._refreshEveryMs
                );
                resolve();
            } catch (error) {
                reject(error);
            }
        });
    }
    public get results(): Record<string, any>[] {
        return [];
    }
}
export default CreateRelationship;

File: flowquery-py/src/parsing/operations/create_relationship.py
"""Represents a CREATE operation for creating virtual relationships."""

from typing import Any, Dict, List, Optional

from ...graph.database import Database
from ...graph.relationship import Relationship
from ..ast_node import ASTNode
from .operation import Operation


class CreateRelationship(Operation):
    """Represents a CREATE operation for creating virtual relationships."""

    def __init__(
        self,
        relationship: Relationship,
        statement: ASTNode,
        is_static: bool = False,
        refresh_every_ms: Optional[int] = None,
    ) -> None:
        super().__init__()
        self._relationship = relationship
        self._statement = statement
        self._is_static = is_static
        self._refresh_every_ms = refresh_every_ms

    @property
    def relationship(self) -> Relationship:
        return self._relationship

    @property
    def statement(self) -> ASTNode:
        return self._statement

    @property
    def is_static(self) -> bool:
        return self._is_static

    @property
    def refresh_every_ms(self) -> Optional[int]:
        return self._refresh_every_ms

    async def run(self) -> None:
        if self._relationship is None:
            raise ValueError("Relationship is null")
        db = Database.get_instance()
        db.add_relationship(
            self._relationship,
            self._statement,
            self._is_static,
            self._refresh_every_ms,
        )

    @property
    def results(self) -> List[Dict[str, Any]]:
        return []
